src/bramble/client/ProjectStats.js

```diff
@@ -0,0 +1,141 @@
+                next();                
+            });
+        
+        }
+
+        // reinitialize project stats
+        _root = root;
+        _cache = {};
+
+        // walk the root
+        _shell.find(_root, { exec:addSize }, function(err){
+            if(err) {
+               return callback(err);
+            }
+            callback();
+        });
+
+    };
+
+    // original unlink
+    var _innerUnlink = _fs.unlink;
+    // overwrite unlink function to do the bookkeeping of project state and call original unlink.
+    _fs.unlink = function(pathname, callback){
+        // call original Filer.unlink
+        _innerUnlink.call(_fs, pathname, function(err){
+            // only update cache once original was successful
+            if(!err){
+                // stop keeping track of deleted file
+                delete _cache[pathname];
+            }
+            
+            // call callback regardless
+            callback(err);
+        });
+    };
+
+    // original writeFile
+    var _innerWriteFile = _fs.writeFile;
+    // overwrite original writeFile and add bookkeeping of project state and call original writeFile.
+    _fs.writeFile = function(filename, data, options, callback){
+        if(typeof options === "function") {
+            callback = options;
+            options = {};
+        }
+
+        _innerWriteFile.call(_fs, filename, data, function(err){
+            if(!err){
+                // owerwrite or add a new size for a given filename
+                _cache[filename] = data.length;
+            }
+            callback(err);
+        });
+    };
+
+    // original rename
+    var _innerRename = _fs.rename;
+    // overwrite original rename and add bookkeeping of project state and call original rename.
+    // this is essential because we don't want to lose track of file's size for renamed files.
+    _fs.rename = function(oldPath, newPath, callback){
+        _innerRename.call(_fs, oldPath, newPath, function(err){
+            // only update cache once original was successful
+            if(!err){
+                _cache[newPath] = _cache[oldPath];
+                delete _cache[oldPath];
+            }
+            
+            // call callback regardless
+            callback(err);
+        });
+    };
+
+    // returns true/false depending on whether root/index.html exists
+    ProjectStats.hasIndexFile = function() {
+        var index = _Path.join(_root, "index.html");
+        // Loop through all paths, return true if one of them is root/index.html
+        return !!(Object.keys(_cache)).find(function(path) {
+            return path === index;
+        });
+    };
+    
+    // returns bytes of files in the root   
+    ProjectStats.getTotalProjectSize = function() {
+        // Sum up all sizes in the cache and return
+        if(_root){
+            return (Object.values(_cache)).reduce(function(acc, val) {
+                return acc + (val ? val : 0);
```

Comment:
Notice how much this code is getting indented?  A good pattern to follow is that you deal with error cases or edge cases first, and return early.  Then you can do your main code and not indent it deep in if statements:

```js
if(!_root) {
  return 0;
}

return (Object.values...
```